onetbb: fix build for ppc/ppc64 and i386

barracuda156 opened the pull request:
#### Description

Updated PR: fixes build for ppc/ppc64 and restores i386.
Closes: https://trac.macports.org/ticket/64936

###### Type(s)
<!-- update (title contains ": U(u)pdate to"), submission (new Portfile) and CVE Identifiers are auto-detected, replace [ ] with [x] to select -->

- [x] bugfix
- [x] enhancement
- [ ] security fix

###### Tested on
<!-- Triple-click and copy the next line and paste it into your shell. It will copy your OS and Xcode version to the clipboard. Paste it here replacing this section.
sh -c 'echo "macOS $(sw_vers -productVersion) $(sw_vers -buildVersion) $(uname -m)"; xcode=$(xcodebuild -version 2>/dev/null); if [ $? == 0 ]; then echo "$(echo "$xcode" | awk '\''NR==1{x=$0}END{print x" "$NF}'\'')"; else echo "Command Line Tools $(pkgutil --pkg-info=com.apple.pkg.CLTools_Executables | awk '\''/version:/ {print $2}'\'')"; fi' | tee /dev/tty | pbcopy
-->
macOS 10.6.8 Server
Xcode 3.2.6

macOS 10.6 PPC (10A190)
Xcode 3.2

macOS 10.5.8
Xcode 3.1.4

###### Verification <!-- (delete not applicable items) -->
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [x] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL? <!-- Please don't open a new Trac ticket if you are submitting a pull request. -->
- [x] checked your Portfile with `port lint --nitpick`?
- [x] tried existing tests with `sudo port test`?
- [ ] tried a full install with `sudo port -vst install`?
- [ ] tested basic functionality of all binary files?

<!-- Use "skip notification" (surrounded with []) to avoid notifying maintainers -->

barracuda156 commented:
@kencu @catap @evanmiller So now it should build on 10.5.8 (ppc, ppc64, universal), 10.6 PPC and 10.6.8 Rosetta. Provisionally i386 should also build. If someone can check that, please update me.

evanmiller commented:
Errors on 10.4

```
:info:build [  1%] Building CXX object src/tbb/CMakeFiles/tbb.dir/arena.cpp.o
:info:build cd /opt/local/var/macports/build/_Users_emiller_macports.local_devel_onetbb/onetbb/work/build/src/tbb && /opt/local/bin/g++-mp-7 -D__TBB_BUILD -D__TBB_USE_ITT_NOTIFY -I/opt/local/var/macports/build/_Users_emiller_macports.local_devel_onetbb/onetbb/work/oneTBB-2021.5.0/src/tbb/../../include -latomic -arch ppc -mmacosx-version-min=10.4 -fPIC -flifetime-dse=1 -Wall -Wextra -Wfatal-errors -D_XOPEN_SOURCE -flto -std=c++11 -MD -MT src/tbb/CMakeFiles/tbb.dir/arena.cpp.o -MF CMakeFiles/tbb.dir/arena.cpp.o.d -o CMakeFiles/tbb.dir/arena.cpp.o -c /opt/local/var/macports/build/_Users_emiller_macports.local_devel_onetbb/onetbb/work/oneTBB-2021.5.0/src/tbb/arena.cpp
:info:build In file included from /opt/local/var/macports/build/_Users_emiller_macports.local_devel_onetbb/onetbb/work/oneTBB-2021.5.0/src/tbb/scheduler_common.h:27:0,
:info:build                  from /opt/local/var/macports/build/_Users_emiller_macports.local_devel_onetbb/onetbb/work/oneTBB-2021.5.0/src/tbb/task_dispatcher.h:24,
:info:build                  from /opt/local/var/macports/build/_Users_emiller_macports.local_devel_onetbb/onetbb/work/oneTBB-2021.5.0/src/tbb/arena.cpp:17:
:info:build /opt/local/var/macports/build/_Users_emiller_macports.local_devel_onetbb/onetbb/work/oneTBB-2021.5.0/src/tbb/co_context.h: In function 'void tbb::detail::r1::create_coroutine(tbb::detail::r1::coroutine_type&, std::size_t, void*)':
:info:build /opt/local/var/macports/build/_Users_emiller_macports.local_devel_onetbb/onetbb/work/oneTBB-2021.5.0/src/tbb/co_context.h:324:11: error: 'getcontext' was not declared in this scope
:info:build      err = getcontext(&c.my_context);
:info:build            ^~~~~~~~~~
:info:build compilation terminated due to -Wfatal-errors.
:info:build make[2]: *** [src/tbb/CMakeFiles/tbb.dir/arena.cpp.o] Error 1
```

You may need to tweak the logic that defines `__TBB_RESUMABLE_TASKS`

barracuda156 commented:
> You may need to tweak the logic that defines `__TBB_RESUMABLE_TASKS`

I am not sure `onetbb` can be supported on 10.4, TBH. But I will look into that.
My original idea was to have `onetbb` on 10.6 and hopefully 10.5.8.

barracuda156 commented:
@mohd-akram @harens I have updated the PR and rebased. Once it builds, please consider merging.
Upstream takes forever to review anything, but we need `oneTBB` working across the board.

mohd-akram commented on devel/onetbb/files/patch-onetbb-older-platforms.diff:
```diff
@@ -0,0 +1,513 @@
+     set(_tbb_target_architectures "${CMAKE_SYSTEM_PROCESSOR}")
+ endif()
+-if ("${_tbb_target_architectures}" MATCHES "(x86_64|amd64|AMD64)") # OSX systems are 64-bit only
++if ("${_tbb_target_architectures}" MATCHES "(i.86|x86|x86_64|amd64|AMD64)") # Clang is not supported on PowerPC
```
This change doesn't seem necessary for PowerPC?

barracuda156 replied:
Yes, but we can support i386 as well.

mohd-akram replied:
Ah okay, I read the PR title (which only mentions PPC) but not the commit message. I guess change the PR title to match.

mohd-akram commented on devel/onetbb/files/patch-onetbb-older-platforms.diff:
```diff
@@ -0,0 +1,513 @@
+-    TBB_EXPORT void __TBB_EXPORTED_FUNC notify_bounded_queue_monitor( concurrent_monitor* monitors, std::size_t monitor_tag
+-                                                            , std::size_t ticket );
++    TBB_EXPORT void __TBB_EXPORTED_FUNC notify_bounded_queue_monitor( concurrent_monitor* monitors, std::size_t monitor_tag,
++                                                            std::size_t ticket );
```
This seems unnecessary. Best to keep the patch minimal to what's required.

barracuda156 replied:
Sure, will remove that now.

mohd-akram commented on devel/onetbb/files/patch-onetbb-older-platforms.diff:
```diff
@@ -0,0 +1,513 @@
+ }; // class task_group_context
+ 
+-static_assert(sizeof(task_group_context) == 128, "Wrong size of task_group_context");
++#if defined __APPLE__ && defined __POWERPC__
```
Use `__ppc__` like above.

barracuda156 replied:
Here and below: whenever ppc32 is meant but not ppc64, `__ppc__` can be used. Otherwise `__POWERPC__` is correct, since it includes both.

mohd-akram commented on devel/onetbb/files/patch-onetbb-older-platforms.diff:
```diff
@@ -0,0 +1,513 @@
+         zone.introspect = &introspect;
+         zone.version = 8;
+         zone.memalign = impl_memalign;
++    #if __MAC_OS_X_VERSION_MAX_ALLOWED >= 1060 && !defined(__POWERPC__) // Unavailable on macOS < 10.6 and any PPC
```
Use `__ppc__`.

mohd-akram commented on devel/onetbb/files/patch-onetbb-older-platforms.diff:
```diff
@@ -0,0 +1,513 @@
+ #elif defined(DOCTEST_PLATFORM_MAC)
+ #if defined(__x86_64) || defined(__x86_64__) || defined(__amd64__) || defined(__i386)
+ #define DOCTEST_BREAK_INTO_DEBUGGER() __asm__("int $3\n" : :) // NOLINT (hicpp-no-assembler)
++#elif defined(__POWERPC__)
```
Use `__ppc__`.

barracuda156 replied:
Here we need `__POWERPC__` (or otherwise `__ppc__ && __ppc64__`, but that’s longer code).

mohd-akram replied:
It seems `__POWERPC__` in caps is [not available on Linux](https://github.com/gcc-mirror/gcc/blob/fe4608efc15b881ac908a3f90d7322736495ae72/gcc/config/rs6000/linux64.h#L292) and other platforms when using GCC. Since you plan to upstream this, it seems it's best to use `__PPC__` when you want to handle both 32-bit and 64-bit.

mohd-akram replied:
> Then we need both, since `__PPC__` is not defined on Apple.

That's what I was worried about. Then I guess either add a check for `__APPLE__` as well to be explicit like in other places, or if this change applies to PPC in general, check for both `__PPC__` and `__POWERPC__`.

barracuda156 replied:
`__POWERPC__` is already Apple-specific (well, it is also defined on BeOS, but nowhere else). I am pretty sure not a single person ever used Macports on BeOS :) 
But we can add a check for Apple too, that won’t hurt.

barracuda156 replied:
You are right though, let’s make it transparent, otherwise this is nowhere obvious.

barracuda156 replied:
@mohd-akram In fact it is fine as-is, as I now noticed (missed that when browsing from mobile): condition is wrapped in `#elif defined(DOCTEST_PLATFORM_MAC)`, so it is already Apple-only, and we do not need to repeat the check.

mohd-akram commented on devel/onetbb/files/patch-onetbb-older-platforms.diff:
```diff
@@ -0,0 +1,513 @@
++#if defined __APPLE__ && defined __POWERPC__
++    // static_assert(sizeof(task_group_context) == 128, "Wrong size of task_group_context");
++#else
++    static_assert(sizeof(task_group_context) == 128, "Wrong size of task_group_context");
```
Negate the condition and wrap this line, like is done in `conformance_allocators.cpp`.

mohd-akram commented on devel/onetbb/files/patch-onetbb-older-platforms.diff:
```diff
@@ -0,0 +1,513 @@
+ }
+ #endif
+-
+-- 
```
Remove these.

mohd-akram commented on devel/onetbb/files/patch-onetbb-older-platforms.diff:
```diff
@@ -0,0 +1,513 @@
+     TestAllocator<oneapi::tbb::cache_aligned_allocator<void>>(Concept);
+     TestAllocator<oneapi::tbb::tbb_allocator<void>>(Concept);
+ 
++  #if !(defined(__APPLE__) && defined(__POWERPC__)) // This test fails to build on Darwin PPC.
```
What's the failure here?

barracuda156 replied:
I need to rerun this. Will update today in a while.

barracuda156 replied:
@mohd-akram Okay, here it is:
```
[ 66%] Building CXX object test/CMakeFiles/conformance_allocators.dir/conformance/conformance_allocators.cpp.o
cd /opt/local/var/macports/build/_opt_PPCRosettaPorts_devel_onetbb/onetbb/work/build/test && /opt/local/bin/g++-mp-11  -I/opt/local/var/macports/build/_opt_PPCRosettaPorts_devel_onetbb/onetbb/work/oneTBB-2021.8.0/test/.. -I/opt/local/var/macports/build/_opt_PPCRosettaPorts_devel_onetbb/onetbb/work/oneTBB-2021.8.0/test -I/opt/local/var/macports/build/_opt_PPCRosettaPorts_devel_onetbb/onetbb/work/oneTBB-2021.8.0/src/tbb/../../include -pipe -Os -Wno-parentheses -Wno-unused-function -DNDEBUG -I/opt/local/include -D_GLIBCXX_USE_CXX11_ABI=0 -arch ppc -mmacosx-version-min=10.6 -Wall -Wextra -Werror -Wfatal-errors -Wshadow -Wcast-qual -Woverloaded-virtual -Wnon-virtual-dtor -D_XOPEN_SOURCE -std=c++11 -MD -MT test/CMakeFiles/conformance_allocators.dir/conformance/conformance_allocators.cpp.o -MF CMakeFiles/conformance_allocators.dir/conformance/conformance_allocators.cpp.o.d -o CMakeFiles/conformance_allocators.dir/conformance/conformance_allocators.cpp.o -c /opt/local/var/macports/build/_opt_PPCRosettaPorts_devel_onetbb/onetbb/work/oneTBB-2021.8.0/test/conformance/conformance_allocators.cpp
/opt/local/var/macports/build/_opt_PPCRosettaPorts_devel_onetbb/onetbb/work/oneTBB-2021.8.0/test/conformance/conformance_allocators.cpp: In function 'void DOCTEST_ANON_FUNC_74()':
/opt/local/var/macports/build/_opt_PPCRosettaPorts_devel_onetbb/onetbb/work/oneTBB-2021.8.0/test/conformance/conformance_allocators.cpp:37:19: error: no matching function for call to 'AssertSameType(std::size_t, std::allocator_traits<tbb::detail::d1::cache_aligned_allocator<int> >::size_type)'
   37 |     AssertSameType(allocator.max_size(), typename std::allocator_traits<Allocator>::size_type(0));
      |     ~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated due to -Wfatal-errors.
make[2]: *** [test/CMakeFiles/conformance_allocators.dir/conformance/conformance_allocators.cpp.o] Error 1
make[2]: Leaving directory `/opt/local/var/macports/build/_opt_PPCRosettaPorts_devel_onetbb/onetbb/work/build'
make[1]: *** [test/CMakeFiles/conformance_allocators.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs...
```
I guess I can open an issue with upstream.

barracuda156 commented on devel/onetbb/files/patch-onetbb-older-platforms.diff:
```diff
@@ -0,0 +1,513 @@
+     TestAllocator<oneapi::tbb::cache_aligned_allocator<void>>(Concept);
+     TestAllocator<oneapi::tbb::tbb_allocator<void>>(Concept);
+ 
++  #if !(defined(__APPLE__) && defined(__POWERPC__)) // This test fails to build on Darwin PPC.
```
Same, since applies to ppc64 too.

barracuda156 commented on devel/onetbb/files/patch-onetbb-older-platforms.diff:
```diff
@@ -0,0 +1,513 @@
+         introspect.force_lock = &zone_force_lock;
+         introspect.force_unlock = &zone_force_unlock;
+         introspect.statistics = zone_statistics;
++    #if __MAC_OS_X_VERSION_MAX_ALLOWED >= 1060 && !defined(__POWERPC__) // zone_locked is not defined on macOS < 10.6 and any PPC
```
Here `__ppc__` gonna work, there is no ppc64 on any existing version of 10.6.x.

mohd-akram replied:
It's fine to leave it as-is then for consistency.

barracuda156 commented:
@mohd-akram I will pause this for a couple of days, to investigate an alternative solution: https://github.com/oneapi-src/oneTBB/pull/840/files#r1100841707
(In any case, we can finalize this very soon.)

barracuda156 commented:
@mohd-akram I think this is ready to be merged, once we see that it builds.

mohd-akram commented on devel/onetbb/files/patch-onetbb-older-platforms.diff:
```diff
@@ -0,0 +1,487 @@
+ ___TBB_malloc_safer_msize
+ ___TBB_malloc_safer_realloc
+ ___TBB_malloc_free_definite_size
++
```
No changes in this file other than this. Should probably be removed.

barracuda156 replied:
Indeed, it is inconsequential and unneeded in our patch.

mohd-akram commented on devel/onetbb/files/patch-onetbb-older-platforms.diff:
```diff
@@ -0,0 +1,487 @@
+     TestAllocator<std::pmr::polymorphic_allocator<void>>(Concept, std::pmr::polymorphic_allocator<void>(&aligned_resource));
+ }
+ #endif
+-
```
Minor nit.

barracuda156 replied:
Gimme a few min, I will rebase.

mohd-akram reviewed (changes requested):
Just a couple of small changes and it's good to merge.

barracuda156 commented:
@mohd-akram Updated the patch, thank you for helping with this!

mohd-akram reviewed (approved):
Thank you!

barracuda156 commented:
@mohd-akram I also plan to work on improving test results, but let us have this merged first.

harens commented:
Thanks @barracuda156 and @mohd-akram! 🎉

barracuda156 commented:
Great, now I will work on tests.
